### PR TITLE
Add JS2Coffee type definitions

### PR DESCRIPTION
**Pull request by ghost**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Comment by SNDST00M** [comment]
@orta @sandersn @RyanCavanaugh any chance you can help debug?

**Comment by sandersn** [comment]
Confusingly, since `@types/babel__core` is on DT, you don't need a package.json at all. But you do need a redirect in tsconfig's compilerOptions:

```
"paths": { "@babel/core": [ "babel__core" ] },
```

**Comment by elibarzilay** [comment]
@rogierschouten -- ping?

**Comment by SNDST00M** [comment]
I accidentally put your name as owner, you'd need to hit <kbd>Approve</kbd on https://github.com/DefinitelyTyped/DefinitelyTyped/pull/55590/files to revert that

File: types/coffeescript/package.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
+    "private": true,
+    "dependencies": {
+        "@babel/core": "^7.1.15"
```
Comment (peterblazejewicz) [inline types/coffeescript/package.json:4]:
```suggestion
        "@babel/core": "~7.7.7"
```
https://github.com/jashkenas/coffeescript/blob/07f644c39223e016aceedd2cd71b5941579b5659/package.json#L43
which one is correct?

Comment (SNDST00M) [reply types/coffeescript/package.json:4]: The one in the `coffeescript` repository would correct, but will it work? `@types/babel__core` is at `7.1.16`

File: types/js2coffee/package.json [hunk 2]
```diff
@@ -0,0 +1,6 @@
+{
+    "private": true,
+    "dependencies": {
+        "source-map": "^0.7.3"
```
Comment (peterblazejewicz) [inline types/js2coffee/package.json:4]: This should use different minor version:
https://github.com/js2coffee/js2coffee/blob/47d3159aba94ffd4377138f7837717d8628eef9f/package.json#L37
But, I think this is ok to use that specific version published `d.ts` types, intead of latest @types/source-map
Am I correct?

File: types/js2coffee/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,396 @@
+// Type definitions for JS2Coffee 2.2
+// Project: https://github.com/js2coffee/js2coffee
+// Definitions by: Munin M. <https://github.com/SNDST00M>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// Minimum TypeScript Version: 4.1
```
Comment (peterblazejewicz) [inline types/js2coffee/index.d.ts:5]: Do you want this specific version? Native JS packages are not most modern JS implementations

File: types/js2coffee/package.json (hunk 2 above)
Comment (SNDST00M) [reply types/js2coffee/package.json:4]: We could go with this version: https://github.com/mozilla/source-map/blob/0.5.7/source-map.d.ts

File: types/js2coffee/index.d.ts (hunk 3 above)
Comment (SNDST00M) [reply types/js2coffee/index.d.ts:5]: https://coffeescript.org/#compatibility I assumed from this that the JS used is a modern lib

**Comment by elibarzilay** [comment]
> I accidentally put your name as owner, you'd need to hit Approve on https://github.com/DefinitelyTyped/DefinitelyTyped/pull/55590/files to revert that

(I'm not sure what you're saying here: did you intend to remove @rogierschouten ?)

File: types/coffeescript/package.json (hunk 1 above)
Comment (peterblazejewicz) [reply types/coffeescript/package.json:4]: Oh, so this one does not need this dependency. I just think @babel/core is outdated on DT (still at 7.1). 
Dependency would resolve to @types/babel__core

**Comment by SNDST00M** [comment]
Yes

File: types/js2coffee/package.json (hunk 2 above)
Comment (peterblazejewicz) [reply types/js2coffee/package.json:4]: Oh, ok. So please review this topic. I'd say, safer is to use @types/source-map@.... as dependency in `package.json` (I understand the source for package is out of DT already, so you need entry in `package.json`, I understand this @types/source-map could require whitelisting, if you decide to use @types/source-map)
I can see now why you've added native packages dependency to `packages.json`, as the original types sources are no longer here

File: types/js2coffee/index.d.ts (hunk 3 above)
Comment (peterblazejewicz) [reply types/js2coffee/index.d.ts:5]: ok, let me rephrase. By putting at 4.1, you're potentially limit this package interoperability with other definitions, as it requires 4.1 version. What's the purpose of this version here?

**Comment by SNDST00M** [comment]
Sorry, tried to make the minimum version of the dependencies match but it failed spectacularly

File: types/js2coffee/package.json (hunk 2 above)
Comment (peterblazejewicz) [reply types/js2coffee/package.json:4]: The removal process was wrong, I think you have to go with 0.7.7 version (sorry for the hassle).
The passghrouh points to distribution which has incorrect ntry in NPM without those types shipped, so
the DT ill never pass a CI (nor install proper deps):
```
npm notice package: source-map@0.5.7
npm notice === Tarball Contents ===
npm notice 2.6kB   package.json
npm notice 23.5kB  README.md
npm notice 1.5kB   LICENSE
npm notice 405B    source-map.js
npm notice 7.9kB   CHANGELOG.md
npm notice 260.2kB dist/source-map.debug.js
npm notice 101.9kB dist/source-map.js
npm notice 26.3kB  dist/source-map.min.js
npm notice 245.8kB dist/source-map.min.js.map
npm notice 3.2kB   lib/array-set.js
npm notice 4.7kB   lib/base64-vlq.js
npm notice 1.5kB   lib/base64.js
npm notice 4.2kB   lib/binary-search.js
npm notice 2.3kB   lib/mapping-list.js
npm notice 3.6kB   lib/quick-sort.js
npm notice 38.4kB  lib/source-map-consumer.js
npm notice 14.1kB  lib/source-map-generator.js
npm notice 13.8kB  lib/source-node.js
npm notice 10.5kB  lib/util.js
npm notice === Tarball Details ===
```

**Comment by peterblazejewicz** [comment]
I've posted explanation why (the removal process was too early for source-map)

**Comment by elibarzilay** [comment]
> Yes, the bot needs them to okay that review

(If you've removed @rogierschouten to avoid waiting for a review then that's wrong.)

File: types/coffeescript/index.d.ts [hunk 4]
```diff
@@ -1,137 +1,9 @@
 // Type definitions for CoffeeScript 2.5
 // Project: https://github.com/jashkenas/coffeescript
-// Definitions by: Munin M. <https://github.com/rogierschouten>
+// Definitions by: Munin M. <https://github.com/SNDST00M>
```
Comment (peterblazejewicz) [inline types/coffeescript/index.d.ts:3]:
```suggestion
// Definitions by: rogierschouten <https://github.com/rogierschouten>
//                 Munin M. <https://github.com/SNDST00M>
```
do you want to retain original author entry? (as the recent discussion)

Comment (SNDST00M) [reply types/coffeescript/index.d.ts:3]: I was the author but copied that by accident - is there any way to have it removed?

Comment (peterblazejewicz) [reply types/coffeescript/index.d.ts:3]: I see, this was from original PR: #54921

Comment (peterblazejewicz) [reply types/coffeescript/index.d.ts:3]: I'd say that's ok to change the header. This was missed on original review (no ack from rogiershouten, so 
now we have this handle in header)

**Comment by peterblazejewicz** [comment]
LGTM
@elibarzilay it looks like simple mistake from initial PR

**Comment by elibarzilay** [comment]
@peterblazejewicz, ah, I missed that

**Comment by SNDST00M** [comment]
Ready to merge
